File: JAVA/Spiral Matrix II.java
public class Solution {
    public int[][] generateMatrix(int n) {
        int cnt = 1;
        int top = 0;
        int bottom = n-1;
        int left = 0;
        int right = n-1;
        int[][] r = new int[n][n];
        while (cnt <= n* n)
        {
            for (int i = left; i <= right; i++)
                r[top][i] = cnt++;
            top++;
            for (int i = top; i <= bottom; i++)
                r[i][right] = cnt++;
            right--;
            for (int i = right; i >= left; i--)
                r[bottom][i] = cnt++;
            bottom--;
            for (int i = bottom; i >= top; i--)
                r[i][left] = cnt++;
            left++;
        }
        return r;
    }
}

File: Spiral Matrix II.cpp
enum Direction { LEFT, RIGHT, UP, DOWN };

class Solution {
public:
    vector<vector<int> > generateMatrix(int n) {
        vector<vector<int> > result(n, vector<int>(n));
        int m = n * n; int num = 1;
        int imin = 0; int imax = n - 1; int jmin = 0; int jmax = n - 1;
        Direction dir = RIGHT;
        while (num <= m) {
            switch (dir) {
            case RIGHT:
                for (int j = jmin; j <= jmax; j++)
                    result[imin][j] = num++;
                imin++; dir = DOWN;
                break;
            case DOWN:
                for (int i = imin; i <= imax; i++)
                    result[i][jmax] = num++;
                jmax--; dir = LEFT;
                break;
            case LEFT:
                for (int j = jmax; j >= jmin; j--)
                    result[imax][j] = num++;
                imax--; dir = UP;
                break;
            case UP:
                for (int i = imax; i >= imin; i--)
                    result[i][jmin] = num++;
                jmin++; dir = RIGHT;
                break;
            }
        }
        return result;
    }
};
